query-service-impl/src/main/java/org/hypertrace/core/query/service/projection/ProjectionTransformation.java

```diff
@@ -341,4 +348,77 @@ private QueryRequest rebuildRequestOmittingDefaults(
         .addAllOrderBy(orderBys)
         .build();
   }
+
+  /*
+   * We need the CONTAINS_KEY filter in all filters and order bys dealing with complex
+   * attribute expressions as Pinot gives error if particular key is absent. Rest all work fine.
+   * To handle order bys, we add the corresponding filter at the top and 'AND' it with the main filter.
+   * To handle filter, we modify each filter (say filter1) as : "CONTAINS_KEY AND filter1".
+   */
+  private Filter rebuildFilterForComplexAttributeExpression(
+      Filter originalFilter, List<OrderByExpression> orderBys) {
+
+    Filter updatedFilter = updateFilterForComplexAttributeExpressionFromFilter(originalFilter);
+    List<Filter> filterList = createFilterForComplexAttributeExpressionFromOrderBy(orderBys);
+
+    if (filterList.isEmpty()) {
```

Comment:
nit: the else clauses here aren't needed and can be unwrapped. Overall, this logic could be written in a more readable way, although I believe it's correct. My suggestion: Start returning `Optional<Filter>` rather than a `Filter` object that may or may not be valid. It better conveys intent and is more readable. The logic can also likely be simplified by combining the top level filters first, then returning a composite if >1, or an optional of the first if not.

Reply:
We are using this convention (of checking with default instance) throughout the code.

Reply:
In new code, we should try to improve that (and go back to fix the old code). Doesn't have to be this PR though.